src/vtok_p11/src/backend/db/object.rs

```diff
@@ -283,6 +284,26 @@ impl Object {
         }
     }
 
+    pub fn new_trusted_cert(info: CertInfo) -> Self {
```

Comment:
TLS termination requires support for Java KeyStore `PrivateKeyEntry` and `TrustedCertificateEntry` only.

No need for `CKO_SECRET_KEY` which correlates to the Java KeyStore [SecretKeyEntry](https://docs.oracle.com/javase/7/docs/api/java/security/KeyStore.SecretKeyEntry.html).

Agreed regarding `CKA_SUBJECT`, this would be used to let the PKCS11 client build a certificate chain by linking each certificate issuer subject name to the next certificate subject name.

When importing the private key along with signed certificate (chain), perhaps it makes sense to create a separate `CKO_CERTIFICATE` object in the database for each certificate in the chain, and specify the `CKA_SUBJECT` attribute to let the certificate chain be reconstructed as required by the PKCS11 client?

Reply:
So `pkcs11` does not specify how a certificate chain would be represented. Each certificate of a chain is represented by a a `CKO_CERTIFICATE` object.

I'm curious if Keystore works correctly and does not mark the certificate as self-signed instead of trusted if only one certificate object is found for the entire chain paired with the private key.

Regardless, bogus PEMs provisioning must be avoided in the current state. Some AWS crypto library foreign function interfaces must be imported (i.e. added in `vtok_p11/src/crypto/ffi.rs`) for parsing the PEM in the backend and extracting the `CKA_SUBJECT` and probably the `CKA_ISSUER` also. These shall implicitly validate the PEM itself.

Looping with `PEM_read_bio_X509()` + `X509_get_subject_name()` until we consume the whole chain and populate with X509 objects (i.e. analogous implementation with `EVP_PKEY`) might do the trick.